File: GO-CODE/RabinKarp.go
package main



func main(){
    text        := "SANDIP"
    pattern     := "NDI"

    // Arrange
    var tl int = len(text)
    var pl int = len(pattern)

    var hashText = 0;
    var hashPat = 0;
    var i,j = 0,0
    var h = 1;
    var primaryNumber = 101
    var d = 256

    for i=0;i<pl -1;i++{
        h = (h * d) % primaryNumber
    }

    for i = 0;i < pl;i++{
        hashPat = (d * hashPat + pattern[i]) % primaryNumber
        hashText = (d * hashText + text[i]) % primaryNumber
    }

    

}

File: PythonCode/RabinKarp.py
class Rabin_Karp:
    def search(self,text,pattern):
        tl = len(text)
        pl = len(pattern)
        i,j,h,d,hashText,hashPatt,primaryNumber = 0,0,1,256,0,0,101

        for i in range(0,len(pattern) - 1):
            h = (h*d) % primaryNumber

        for i in range(0,len(pattern) - 1):
            hashText = (d * hashText + ord(text[i])) % primaryNumber
            hashPatt = (d * hashPatt + ord(pattern[i])) % primaryNumber

        print(hashText,hashPatt)
        for i in range(0, tl - pl):
            k = 0
            if hashPatt == hashText:

                for j in range(0, pl - 1):
                    if text[ i + j] != pattern[j]:
                        break
                    k = k + 1

                if k == pl-1:
                    print("here")
                    print("Match found at "+str(i))


text = "RED BROWN FOX JUMPED ON ROCK"
pattern= "REDI"

obj = Rabin_Karp()
obj.search(text,pattern)
